rasa/shared/core/domain.py

```diff
@@ -252,195 +255,213 @@ def from_directory(cls, path: Text) -> "Domain":
-                domain_dict[KEY_ACTIONS].remove(form)
+        return combined, duplicates
 
+    @staticmethod
+    def _merge_actions_and_e2e_actions(
+        combined: Dict, domain_dict: Dict, duplicates: Dict
+    ) -> Tuple:
         for key in [KEY_ACTIONS, KEY_E2E_ACTIONS]:
-            combined[key] = self.merge_lists(combined[key], domain_dict[key])
+            duplicates[key] = rasa.shared.utils.common.extract_duplicates(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
+            combined[key] = rasa.shared.utils.common.merge_lists(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
 
-        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
-            combined[key] = self.merge_dicts(combined[key], domain_dict[key], override)
+        return combined, duplicates
 
-        return self.__class__.from_dict(combined)
+    @staticmethod
+    def _merge_forms_responses_slots(
+        combined: Dict, domain_dict: Dict, duplicates: Dict, override: bool = False
+    ) -> Tuple:
+        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
+            duplicates[key] = rasa.shared.utils.common.extract_duplicates(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
+            combined[key] = rasa.shared.utils.common.merge_dicts(
+                combined.get(key, {}), domain_dict.get(key, {}), override
+            )
+        return combined, duplicates
 
+    @staticmethod
     def merge_domain_dicts(
-        self, domain1: Dict, domain2: Dict, override: bool = False
-    ) -> Dict[Text, Any]:
-        """Merges this domain dict with another one, combining their attributes.
+        domain_dict: Dict, combined: Dict, override: bool = False, is_dir: bool = False
+    ) -> Dict:
+        """Combines two domain dictionaries."""
+        if not domain_dict:
+            return combined
 
-        This is used when multiple domain yml files are configured in a single
-        directory. Unlike the merge method above, which merges Domain objects by
-        creating each object then merging it with the previous, this method merges
-        domain dicts, and ensures all attributes (like ``intents``, ``entities``, and
-        ``actions``) are known to the Domain when the object is created.
+        if not combined:
+            return domain_dict
 
-        List attributes like ``intents`` and ``actions`` are deduped
-        and merged. Single attributes are taken from `domain1` unless
-        override is `True`, in which case they are taken from `domain2`.
-        """
-        if not domain2:
-            return domain1
-
-        if not domain1:
-            return domain2
-
-        domain_dict = domain2
-        combined = domain1
+        if not combined.get(KEY_TRAINING_DATA_FORMAT_VERSION):
+            combined[
+                KEY_TRAINING_DATA_FORMAT_VERSION
+            ] = LATEST_TRAINING_DATA_FORMAT_VERSION
```

Comment:
Don't we do this also in `_add_default_keys_to_domain_dict`?

Reply:
Ah yes, this now becomes redundant. It was in the original code before any PR changes, I will remove it.